File: DataStructure/01_Arrays/Concepts/Javascript/04_mergeSort.js
function mergeSort(arr) {
  if (arr.length <= 1) return arr; // Base case: already sorted
  let mid = Math.floor(arr.length / 2);
  let left = mergeSort(arr.slice(0, mid)); // Sort left half
  let right = mergeSort(arr.slice(mid)); // Sort right half
  return merge(left, right);
}

function merge(left, right) {
  let sortedArray = [],
    i = 0,
    j = 0;
  while (i < left.length && j < right.length) {
    if (left[i] < right[j]) {
      sortedArray.push(left[i]);
      i++;
    } else {
      sortedArray.push(right[j]);
      j++;
    }
  }
  // Append remaining elements
  return [...sortedArray, ...left.slice(i), ...right.slice(j)];
}
console.log(mergeSort([64, 34, 25, 12, 22, 11, 90]));

File: DataStructure/01_Arrays/Concepts/Python/04_mergeSort.py
def merge_sort (arr):
    if len(arr)<=1:
        return arr
    mid = len(arr) // 2
    left = merge_sort(arr[:mid])
    right = merge_sort(arr[mid:])
    return merge(left,right)

def merge(left, right):
    sorted_array = []
    i=j=0
    while i < len(left) and j < len(right):
        if left[i] < right[j]:
            sorted_array.append(left[i])
            i += 1
        else:
            sorted_array.append(right[j])
            j += 1
    return sorted_array + left[i:] + right[j:]

print(merge_sort([64, 34, 25, 12, 22, 11, 90]))
